File: project_euler/problem_201/solution.py
"""
Problem 201: Subsets with a Unique Sum

For S = {1^2, 2^2, ..., 100^2}, find the sum of all integers that are the
sum of exactly one subset of S of size 50.

For each subset size we maintain two bitsets:
- once[j]: sums achievable in exactly one way
- many[j]: sums achievable in at least two ways
"""


def solve():
    n = 100
    k = 50
    elements = [i * i for i in range(1, n + 1)]

    prefix = [0]
    for value in elements:
        prefix.append(prefix[-1] + value)

    min_sum = prefix[k]
    max_sum = prefix[n] - prefix[n - k]
    mask = (1 << (max_sum + 1)) - 1

    once = [0] * (k + 1)
    many = [0] * (k + 1)
    once[0] = 1  # Only sum 0 is achievable with 0 elements.

    for value in elements:
        for size in range(k, 0, -1):
            generated_once = (once[size - 1] << value) & mask
            generated_many = (many[size - 1] << value) & mask

            new_many = many[size] | generated_many | (once[size] & generated_once)
            new_once = (once[size] ^ generated_once) & ~new_many

            many[size] = new_many
            once[size] = new_once & mask

    answer = 0
    bits = once[k]
    for total in range(min_sum, max_sum + 1):
        if (bits >> total) & 1:
            answer += total

    print(answer)


if __name__ == "__main__":
    solve()

File: project_euler/problem_201/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int N = 100;
    const int K = 50;

    vector<int> squares(N);
    for (int i = 0; i < N; i++) {
        squares[i] = (i + 1) * (i + 1);
    }

    vector<int> prefix(N + 1, 0);
    for (int i = 0; i < N; i++) {
        prefix[i + 1] = prefix[i] + squares[i];
    }

    int minSum = prefix[K];
    int maxSum = prefix[N] - prefix[N - K];

    // dp[j][s] is 0, 1, or 2 depending on whether there are
    // zero, one, or at least two j-element subsets with sum s.
    vector<vector<uint8_t>> dp(K + 1, vector<uint8_t>(maxSum + 1, 0));
    dp[0][0] = 1;

    for (int i = 1; i <= N; i++) {
        int value = squares[i - 1];
        int previousTotal = prefix[i - 1];

        for (int size = min(i, K); size >= 1; size--) {
            int minPrev = prefix[size - 1];
            int maxPrev = previousTotal - prefix[i - size];
            auto& curr = dp[size];
            const auto& prev = dp[size - 1];

            for (int prevSum = maxPrev; prevSum >= minPrev; prevSum--) {
                if (prev[prevSum]) {
                    int newSum = prevSum + value;
                    int total = curr[newSum] + prev[prevSum];
                    curr[newSum] = static_cast<uint8_t>(min(total, 2));
                }
            }
        }
    }

    long long answer = 0;
    for (int sum = minSum; sum <= maxSum; sum++) {
        if (dp[K][sum] == 1) {
            answer += sum;
        }
    }

    cout << answer << '\n';
    return 0;
}
